controllers/pipelinegates/pipelinegate_controller.go

```diff
@@ -0,0 +1,376 @@
+			log.Error(err, "Failed to update PipelineGate status at CR")
+			return ctrl.Result{}, err
+		}
+
+		log.V(2).Info(fmt.Sprintf("Updated gate state lastReported to %s.", string(lastReported)), "Name", gate.Name, "ID", gate.Spec.ID, "State", *gate.Status.State, "LastReportedAt", *gate.Status.LastReportedAt, "LastSyncedAt", gate.Spec.LastSyncedAt, "jobRef", *gate.Status.JobRef)
+
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func hasFailed(job *batchv1.Job) (bool, *batchv1.JobCondition) {
+	conditions := job.Status.Conditions
+	// check if the conditions slice contains a failed condition
+	// failed means the backoff limit was reached and it is not being retried anymore!
+	for _, condition := range conditions {
+		if condition.Type == batchv1.JobFailed {
+			return true, &condition
+		}
+	}
+	return false, nil
+}
+
+func hasSucceeded(job *batchv1.Job) (bool, *batchv1.JobCondition) {
+	conditions := job.Status.Conditions
+	// check if the conditions slice contains a failed condition
+	for _, condition := range conditions {
+		if condition.Type == batchv1.JobComplete {
+			return true, &condition
+		}
+	}
+	return false, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PipelineGateReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&pipelinesv1alpha1.PipelineGate{}).
+		Owns(&batchv1.Job{}).
+		Complete(r)
+
+}
+
+// Job reconciles a k8s job object.
+func Job(ctx context.Context, r client.Client, job *batchv1.Job, log logr.Logger) (*batchv1.Job, error) {
+	foundJob := &batchv1.Job{}
+	justCreated := false
+	if err := r.Get(ctx, types.NamespacedName{Name: job.Name, Namespace: job.Namespace}, foundJob); err != nil {
+		if apierrs.IsNotFound(err) {
+			log.V(2).Info("Creating Job.", "Namespace", job.Namespace, "Name", job.Name)
+			if err := r.Create(ctx, job); err != nil {
+				log.Error(err, "Unable to create Job.")
+				return nil, err
+			}
+			justCreated = true
+		} else {
+			log.Error(err, "Error getting Job.")
+			return nil, err
+		}
+	}
+	if !justCreated && CopyJobFields(job, foundJob, log) {
+		log.V(2).Info("Updating Job.", "Namespace", job.Namespace, "Name", job.Name)
+		if err := r.Update(ctx, foundJob); err != nil {
+			if apierrs.IsConflict(err) {
+				return foundJob, nil
+			}
+
+			log.Error(err, "Unable to update Job")
+			return foundJob, err
+		}
+	}
+	if justCreated {
+		return job, nil
+	}
+	return foundJob, nil
+}
+
+func (r *PipelineGateReconciler) generateJob(ctx context.Context, log logr.Logger, jobSpec batchv1.JobSpec, jobRef console.NamespacedName) *batchv1.Job {
+	return &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      jobRef.Name,
+			Namespace: jobRef.Namespace,
+		},
+		Spec: jobSpec,
+	}
+}
+
+func CopyJobFields(from, to *batchv1.Job, log logr.Logger) bool {
+	requireUpdate := false
+	if !reflect.DeepEqual(to.Spec.Template.Spec.Volumes, from.Spec.Template.Spec.Volumes) {
```

Comment:
Can we just copy whole spec.template.spec? Then we can just log that spec has changed. This seems to be a bit too detailed.